core/analysis/embedding_classification_stream.cpp

```diff
@@ -0,0 +1,238 @@
+    IR_FRMT_ERROR(
+      "Caught error while normalizing text_token_normalizing_stream from JSON");
+  }
+  return false;
+}
+
+REGISTER_ANALYZER_VPACK(irs::analysis::embedding_classification_stream, make_vpack, normalize_vpack_config);
+REGISTER_ANALYZER_JSON(irs::analysis::embedding_classification_stream, make_json, normalize_json_config);
+
+} // namespace
+
+namespace iresearch {
+namespace analysis {
+
+std::shared_ptr<fasttext::FastText> EmbeddingsModelLoader::get_model_and_increment_count(const std::string& model_location) {
+  std::unique_lock<std::mutex> lock(this->map_mutex);
+  if (model_map.find(model_location) == this->model_map.end()) {
+    auto ft = iresearch::memory::make_shared<fasttext::FastText>();
+    ft->loadModel(model_location);
+    this->model_map[model_location] = std::move(ft);
+    this->model_usage_count[model_location] = 0;
+  }
+  this->model_usage_count[model_location] += 1;
+  return this->model_map.at(model_location);
+}
+
+void EmbeddingsModelLoader::decrement_model_usage_count(const std::string& model_location) {
+  std::unique_lock<std::mutex> lock(this->map_mutex);
+  if (this->model_map.find(model_location) == this->model_map.end()) {
+    // Something's gone horribly wrong here.
+    // TODO: Throw exception to escape invalid state
+  } else {
+    this->model_usage_count[model_location] -= 1;
+    if (this->model_usage_count[model_location] == 0) {
+      this->model_usage_count.erase(model_location);
+      this->model_map.erase(model_location);
+    }
+  }
+}
+
+embedding_classification_stream::embedding_classification_stream(const Options &options)
+  : analyzer{irs::type<embedding_classification_stream>::get()},
+  model_container{EmbeddingsModelLoader::getInstance().get_model_and_increment_count(options.model_location)},
+  model_location{options.model_location} {}
+
+
+embedding_classification_stream::~embedding_classification_stream() {
+  EmbeddingsModelLoader::getInstance().decrement_model_usage_count(model_location);
+}
+
+void embedding_classification_stream::init() {
+  REGISTER_ANALYZER_JSON(embedding_classification_stream, make_json, normalize_json_config);
+  REGISTER_ANALYZER_VPACK(embedding_classification_stream, make_vpack, normalize_vpack_config);
+}
+
+bool embedding_classification_stream::next() {
+  if (term_eof_) {
+    return false;
+  }
+
+  term_eof_ = true;
+
+  return true;
+}
+
+bool embedding_classification_stream::reset(const string_ref& data) {
+  // convert encoding to UTF-8
+  auto& term = std::get<term_attribute>(attrs_);
+
+  term.value = bytes_ref::NIL; // clear the term value
+  term_eof_ = true;
+
+  auto& offset = std::get<irs::offset>(attrs_);
+  offset.start = 0;
+  offset.end = static_cast<uint32_t>(data.size());
+
+  std::get<payload>(attrs_).value = ref_cast<uint8_t>(data);
+  term_eof_ = false;
+
+  // Now classify token!
+  std::vector<std::pair<float, std::string>> predictions{};
+  std::string s{data.c_str(), data.size()};
+  std::stringstream ss{s};
```

Comment:
we can avoid copying input data by supplying a custom implementation of `rdbuf` to  `std::istream`. 
Check out `data_input.hpp:155`. The idea is to simply wrap input with `bytes_ref_input` and `input_buf`. To avoid unnecessary virtual calls the latter can be even turned into a template